doc/contributing.md

```diff
@@ -19,14 +12,225 @@ To make a code contribution to the project:
+
+### Adding new features
+
+This section will help you get started adding new features to SkyPortal. We'll walk through adding a new feature: a button that triggers an API call to the back-end, which generates and returns a random string that is stored in the front-end redux state, triggering an automatic re-render of the React component that displays that value in our example component.
+
+#### Front-end: React & Redux
+
+The SkyPortal front-end uses React in conjunction with Redux for state management. See the following links for introductions to these libraries:
+
+https://reactjs.org/tutorial/tutorial.html
+https://reactjs.org/docs/hooks-intro.html
+https://redux.js.org/basics/basic-tutorial/
+
+We'll start with defining our new `RandomString` component, which accesses (reads, not modifies) the Redux state via the `useSelector` hook and dispatches action creators with the `useDispatch` hook (see tutorials links) in a new file in the components directory (`static/js/components/RandomString.js`):
+
+``` javascript
+import React from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+
+// Import our actions from `static/js/ducks/randomString.js`
+import * as Actions from '../ducks/randomString';
+
+
+const RandomString = () => {
+  const value = useSelector((state) => state.randomString);
+  const dispatch = useDispatch();
+
+  return (
+    <div>
+      <div>
+        Back-end random string generator.
+        <br />
+        This is a dummy component for educational purposes.
+        <br />
+        Current value is: {value}
+      </div>
+      <div>
+        <button type="button" onClick={() => dispatch(Actions.fetchRandomString())}>
+          Generate new random string on the back-end.
+        </button>
+      </div>
+    </div>
+  );
+};
+
+export default RandomString;
+```
+
+All of the redux-related logic associated with our new component will live in a separate file in `static/js/ducks`. This is where we'll define our action creators and reducer(s) (`static/js/ducks/randomString.js)`:
+
+``` javascript
+import * as API from '../API';
+
+export const FETCH_RANDOM_STRING = 'skyportal/FETCH_RANDOM_STRING';
+export const FETCH_RANDOM_STRING_OK = 'skyportal/FETCH_RANDOM_STRING_OK';
+const FETCH_RANDOM_STRING_FAIL = 'skyportal/FETCH_RANDOM_STRING_FAIL';
+
+export function fetchRandomString() {
+  return API.GET('/api/random_string', FETCH_RANDOM_STRING);
+}
+
+export default function reducer(state="null", action) {
+  switch (action.type) {
+    case FETCH_RANDOM_STRING_OK: {
+      const { value } = action.data;
+      return value;
+    }
+    case FETCH_RANDOM_STRING_FAIL: {
+      return "null";
+    }
+    default:
+      return state;
+  }
+}
+```
+
+And we'll render our new `RandomString` component at the bottom of our profile page. The `Profile` component is defined in `static/js/components/Profile.jsx`:
+
+``` javascript
```

Comment:
Maybe just show the relevant sections of that file, using `...` to hide other lines that the user shouldn't care about.